firedrake/slate/slac/compiler.py

```diff
@@ -530,6 +590,325 @@ def parenthesize(arg, prec=None, parent=None):
+                    PetscScalar* Awork=(PetscScalar*) malloc(N*N*sizeof(PetscScalar));
+                    LAPACKgetrf_(&N,&N,A,&N,ipiv,&info);
+                    if(info==0)
+                        LAPACKgetri_(&N,A,&N,ipiv,Awork,&N,&info);
+                    if(info!=0)
+                        fprintf(stderr,\"Getri throws nonzero info.\");
+                }
+                #endif
+            """
+            yield("lapack_inverse", "#include <petscsystypes.h>\n#include <petscblaslapack.h>\n"+inverse_preamble)
+            return
+
+    def inv_fn_lookup(target, identifier):
+        if identifier == 'inv':
+            return INVCallable(name='inv')
+
+        return None
+
+    loopy_merged = loopy.register_function_id_to_in_knl_callable_mapper(loopy_merged, inv_fn_lookup)
+
+    return loopy_merged
+
+
+def get_solve_callable(loopy_merged):
+    class SolveCallable(loopy.ScalarCallable):
+        def __init__(self, name, arg_id_to_dtype=None, 
+                     arg_id_to_descr=None, name_in_target=None):
+
+            super(SolveCallable, self).__init__(name,
+                  arg_id_to_dtype=arg_id_to_dtype,
+                  arg_id_to_descr=arg_id_to_descr)
+
+            self.name = name
+            self.name_in_target = name_in_target
+            
+        def with_types(self, arg_id_to_dtype, kernel, callables_table):
+            for i in range(0, len(arg_id_to_dtype)):
+                if i not in arg_id_to_dtype or arg_id_to_dtype[i] is None:
+                    # the types provided aren't mature enough to specialize the
+                    # callable
+                    return (self.copy(arg_id_to_dtype=arg_id_to_dtype),
+                            callables_table)
+
+            mat_dtype = arg_id_to_dtype[0].numpy_dtype
+            name_in_target = "solve_"
+
+            from loopy.types import NumpyType
+            return (self.copy(name_in_target=name_in_target,
+                              arg_id_to_dtype={0: NumpyType(mat_dtype), 1: NumpyType(mat_dtype), 2: NumpyType(int)}),
+                    callables_table)
+
+        def emit_call_insn(self, insn, target, expression_to_code_mapper):
+            assert self.is_ready_for_codegen()
+            assert isinstance(insn, loopy.CallInstruction)  # for batched this should be call instruction
+
+            parameters = insn.expression.parameters
+
+            if type(parameters) != list:
+                parameters = list(parameters)
+            par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]  # TODO: get the reads right
+
+            parameters.append(insn.assignees[0])
+            par_dtypes.append(self.arg_id_to_dtype[0])
+
+            from loopy.expression import dtype_to_type_context
+            from pymbolic.mapper.stringifier import PREC_NONE
+            from loopy.symbolic import SubArrayRef
+            from pymbolic import var
```

Comment:
local imports

Reply:
Adressed in 7aa9b698a1722b7f18f5b3f23332dbb95f064226